msg/async/rdma: add iWARP RDMA protocol support

tanghaodong25 opened the pull request:
Add RDMA communication management with librdmacm to support iWARP protocol. 

Signed-off-by: Haodong Tang <haodong.tang@intel.com>

tanghaodong25 commented:
@yuyuyu101 Please help to review. 

1. Add two classes, RDMACMConnectedSocketImpl and RDMACMServerSocketImpl, which are inherited from RDMAConnectedSocketImpl and RDMAServerSocketImpl. We overwrote the connect/accept/shutdown/close method of supper class.
2. Use ibv_context* as the identity for every RDMA device in class Infiniband. 

We would like to hear more your advice, besides, we would have more test based on this patch with RoCE V2 and iWARP. 

Thanks.

tanghaodong25 commented:
Attached the initial 4K random write performance result with iWARP protocol. In general, we got up to 8% performance gain with RDMA compared with TCP/IP.  We are expanding the Ceph scale and get more performance result. 

### The configuration:
#### Hardware
1x OSD node, 1x OSD per SSD. Intel Purley CPU, 128 GB Memory, 4x NVMe as OSD drive, 1x Optane as BlueStore DB drive. 
#### Software
CentOS7

![iwarp_perf](https://user-images.githubusercontent.com/5491410/35785428-541e7d3e-0a5b-11e8-80ad-65d1e8e7683a.png)

amitkumar50 commented on src/msg/async/rdma/Infiniband.cc:
```diff
@@ -190,20 +196,32 @@ int Infiniband::QueuePair::init()
+    qp = cm_id->qp;
   }
-
+  
```
NIT;tab

tanghaodong25 replied:
Fixed, thanks. I'll update later.

yuyuyu101 commented:
looks reasonable

majianpeng commented on src/msg/async/rdma/RDMACMConnectedSocketImpl.cc:
```diff
@@ -0,0 +1,155 @@
+        lderr(cct) << __func__ << " failed to create qp" << dendl; 
+        notify();
+      }
+      local_qpn = qp->get_local_qp_number();
```
how to handle if qp == NULL?

tanghaodong25 replied:
@majianpeng fixed, thanks.

tchaikov commented on src/msg/async/rdma/RDMAStack.h:
```diff
@@ -237,25 +246,72 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
+    virtual int listen(entity_addr_t &sa, const SocketOptions &opt) override;
+    virtual int accept(ConnectedSocket *s, const SocketOptions &opts, entity_addr_t *out, Worker *w) override;
+  private:
+    struct rdma_cm_id *cm_id;
```
nit, unlike C lanuage, no need to add `struct` before the struct type name.

tanghaodong25 replied:
@tchaikov fixed, updated the code, thanks.

tchaikov commented on src/msg/async/rdma/RDMAStack.h:
```diff
@@ -164,13 +164,21 @@ class RDMAWorker : public Worker {
   typedef Infiniband::CompletionQueue CompletionQueue;
-
- private:
+  
```
trailing space.

tchaikov commented on src/msg/async/rdma/RDMACMConnectedSocketImpl.cc:
```diff
@@ -0,0 +1,155 @@
+      local_qpn = qp->get_local_qp_number();
+      my_msg.qpn = local_qpn;
+      
+      cm_params.retry_count = 7;
```
magic number.

tchaikov commented on src/msg/async/rdma/RDMAServerSocketImpl.cc:
```diff
@@ -75,8 +75,9 @@ int RDMAServerSocketImpl::listen(entity_addr_t &sa, const SocketOptions &opt)
 {
   ldout(cct, 15) << __func__ << dendl;
-
+  
```
trailing space

tanghaodong25 commented:
@yuyuyu101 rebased the code as we talked at the CephAlocon a few days ago, please help to review.

The patch includes two PRs:
- Introduced the RDMA-CM library with the first PR.
- Added "ibv_post_recv/send" support with the second PR.

This patch was for iWAPR RDMA protocol.  Added "ms_async_rdma_type" to specify RDMA protocol type, that could be IWARP, RoCE or IB. Since IWARP needs RDMA-CM library, the "ms_async_rdma_cm" was to control if we'd use RDMA-CM or self-implemented network connection management.

yuyuyu101 commented on src/common/legacy_config_opts.h:
```diff
@@ -165,13 +165,19 @@ OPTION(ms_async_rdma_send_buffers, OPT_U32)
 
+// rdma connection management
+OPTION(ms_async_rdma_cm, OPT_BOOL)
+OPTION(ms_async_rdma_type, OPT_STR)
```
I don't think we need two options? ib, roce, iwarp is enough. currently roce can run into verbs libs because original rdma-cm can't work

tanghaodong25 replied:
I suppose RoCE V1 works well with current implementation, but RoCE V2 needs rdma-cm as well?

yuyuyu101 commented on src/msg/async/rdma/Infiniband.cc:
```diff
@@ -890,7 +913,11 @@ void Infiniband::init()
   assert(NetHandler(cct).set_nonblock(device->ctxt->async_fd) == 0);
 
-  rx_queue_len = device->device_attr->max_srq_wr;
+  support_srq = cct->_conf->ms_async_rdma_support_srq;
```
make srq support related to rdma type?

tanghaodong25 replied:
Current iWARP NIC doesn't support SRQ, but it should support it in the near future.

yuyuyu101 commented:
plz rebase

tanghaodong25 commented:
@yuyuyu101 rebased.

yuyuyu101 commented:
do you tested this with roce?

tanghaodong25 commented:
RoCE V2 works fine with this patch. I haven't tested for RoCE V1.

yuyuyu101 commented:
I don't have iwarp adapter currently, but I will borrow several cards from chelisio and I think I will test this later. the pr overall is ok to me. I think we can merge this firstly since it has stuck too long

tanghaodong25 commented:
Great, thank you Haomai.

tchaikov commented on cmake/modules/Findrdma.cmake:
```diff
@@ -10,16 +10,20 @@ find_path(RDMA_INCLUDE_DIR infiniband/verbs.h)
 
 if (RDMA_FOUND)
   message(STATUS "Found libibverbs: ${RDMA_LIBRARY}")
+  message(STATUS "Found librdmacm: ${RDMACM_LIBRARY}")
```
please note, if `RDMACM_LIBRARY ` is not found. `find_package(rdma REQUIRED)` won't fail. i'd suggest move `rdmacm` detection into its own `FindRDMAcm.cmake` module.

tanghaodong25 replied:
Fixed. Please review.

tanghaodong25 commented:
@tchaikov 
Hi Kefu, as your suggestion, I added Findrdmacm.cmake to detect librdmacm, please help to review.

tanghaodong25 commented on cmake/modules/Findrdmacm.cmake:
```diff
@@ -0,0 +1,39 @@
+find_library(RDMACM_LIBRARY NAMES ${RDMACM_NAMES})
+
+if (RDMACM_INCLUDE_DIR AND RDMACM_LIBRARY)
+  set(RDMACM_FOUND TRUE)
```
Fixed, please take a look.

tanghaodong25 commented on src/msg/async/rdma/RDMAStack.h:
```diff
@@ -237,25 +246,87 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
+class RDMAIWARPServerSocketImpl : public RDMAServerSocketImpl {
+  public:
+    RDMAIWARPServerSocketImpl(CephContext *cct, Infiniband *i, RDMADispatcher *s, RDMAWorker *w, entity_addr_t& a);
+    virtual int listen(entity_addr_t &sa, const SocketOptions &opt) override;
```
@tchaikov  Function listen is used at this [line](https://github.com/ceph/ceph/pull/20297/commits/74a63c59c98a8741f39b13cbad5d54d2aaf2415f#diff-028b4fcde35bfcf6b3e1e5473b46ce61R499). It should be prefixed with virtual?

tanghaodong25 commented on src/msg/async/rdma/RDMAConnectedSocketImpl.cc:
```diff
@@ -13,30 +13,31 @@
-  notify_fd = dispatcher->register_qp(qp, this);
-  dispatcher->perf_logger->inc(l_msgr_rdma_created_queue_pair);
-  dispatcher->perf_logger->inc(l_msgr_rdma_active_queue_pair);
+  if (!cct->_conf->ms_async_rdma_cm) {
```
@tchaikov actually, ```info``` is not used in RDMAConnectedSocketImpl, the info for rdmacm is used in Class RDMAIWARPConnectedSocketImpl, so I deleted this  argument.

tanghaodong25 commented:
@tchaikov @yuyuyu101 Do you have any follow-up reviews or suggestions on this patch? Feel free to let me know, I will fix it as soon as possible.

tchaikov commented on cmake/modules/Findrdmacm.cmake:
```diff
@@ -0,0 +1,28 @@
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(rdmacm DEFAULT_MSG RDMACM_LIBRARIES RDMACM_INCLUDE_DIR)
+
+if(RDMACM_FOUND)
```
again, please remove this `if` block. it reinvents `find_package_handle_standard_args()`.

tchaikov commented on cmake/modules/Findrdmacm.cmake:
```diff
@@ -0,0 +1,28 @@
+endif()
+
+mark_as_advanced(
+  RDMACM_LIBRARY
```
`RDMACM_LIBRARY` is set nowhere.

tchaikov commented on src/CMakeLists.txt:
```diff
@@ -789,8 +800,8 @@ endif (WITH_MGR)
 add_executable(librados-config ${librados_config_srcs})
-target_link_libraries(librados-config librados global ${BLKID_LIBRARIES} ${RDMA_LIBRARIES}
-  ${CMAKE_DL_LIBS})
+target_link_libraries(librados-config librados global ${BLKID_LIBRARIES} ${VERBS_LIBRARIES}
```
i don't think we should link against RDMA_LIBRARIES or VERBS_LIBRARIES directly here. as this linkage is introduced by libcommon, and librados-config should not use symbols offered by these two libs.

tchaikov commented on src/CMakeLists.txt:
```diff
@@ -899,7 +910,7 @@ set(ceph_osd_srcs
 add_dependencies(ceph-osd erasure_code_plugins)
 target_link_libraries(ceph-osd osd os global-static ceph-common
-  ${BLKID_LIBRARIES} ${RDMA_LIBRARIES})
+  ${BLKID_LIBRARIES} ${VERBS_LIBRARIES} ${RDMACM_LIBRARIES})
```
ditto.

tchaikov commented on src/msg/async/rdma/RDMAIWARPConnectedSocketImpl.cc:
```diff
@@ -0,0 +1,182 @@
+  ldout(cct, 20) << __func__ << " event name: " << rdma_event_str(event->event)
+                             << " (cm id: " << cm_id << ")" << dendl;
+  struct rdma_conn_param cm_params;
+  switch(event->event) {
```
add a space after `switch`.

tchaikov commented:
http://pulpito.ceph.com/kchai-2018-06-27_11:05:15-rados-wip-kefu-testing-2018-06-27-1254-distro-basic-smithi/

LenzGr commented:
Just stumbled over this PR after it was merged. Doesn't this feature require some form of documentation updates?

tanghaodong25 commented:
@LenzGr I'll add the document later.

qlogicbj commented:
@tanghaodong25 Do you update the document/guide ?
